### PR TITLE
Accept HTTPoison options, upgrade :goth package

### PR DESCRIPTION
**Pull request by nitinstp23**
**Changes**

* Use `jason` instead of `poison` for JSON encoding/decoding
* Upgrade `:goth` to `0.11.1`, as they have added support to disable `:goth` in other environments
* Accept `HTTPoison` and `:hackney` options via `default_options` setting -
```elixir
config :goth,
  default_options: [
    timeout: 10_000,
    recv_timeout: 10_000,
    hackney: [pool: :firebase_hackney_pool]
  ]
```

**Breaking Changes**

* Pass `client_email` of each firebase project for `Auth` & `Messaging` APIs
   Goth now requires `client_email` to fetch the `oauth_token`.

### CODE REVIEW

**Comment by nitinstp23** [comment]
@bernardd Can you review this PR?

**Comment by bernardd** [comment]
I can, but not until I'm back in the office next week.

**Comment by bernardd** [comment]
Hey Nitin,
So I've got a few questions:
* Is there any particular reason for moving from `Poison` to `Jason`? I don't have a preference, I'm just curious.

* Looking at the latest `goth` code, it looks like the client email should be supplied as part of a tuple, and passed in here: https://github.com/peburrows/goth/blob/94fcb6dedbc246da122576e429a47a53f373a210/lib/goth/token.ex#L78
That means that this line:
```elixir
{:ok, token} = Goth.Token.for_scope(client_email, @auth_scope)
```
should actually be
```elixir
{:ok, token} = Goth.Token.for_scope({client_email, @auth_scope})
```
* Also, since it seems likely that a common use case will have only one Firebase account, forcing the user to supply the `client_email` parameter for all calls seems a bit unweildy. My inclination would be to make it an optional parameter. That is:
```elixir
get_user(key, value, client_email)
```
becomes
```elixir
get_user(key, value, client_email \\ nil)
```
(and similarly for other functions). Then `auth_header()` would look like:

```elixir
  defp auth_header(nil) do
    {:ok, token} = Goth.Token.for_scope(@auth_scope)

  defp auth_header(client_email) do
    {:ok, token} = Goth.Token.for_scope({client_email, @auth_scope})
```

**Comment by nitinstp23** [comment]
Hey @bernardd,

> Is there any particular reason for moving from Poison to Jason? I don't have a preference, I'm just curious.

The reason for moving to Jason is performance, its almost twice as fast as `Poison` - https://gist.github.com/michalmuskala/4d64a5a7696ca84ac7c169a0206640d5

---------------------------------------------------------------------------------------------

> Looking at the latest goth code, it looks like the client email should be supplied as part of a tuple, and passed in here: https://github.com/peburrows/goth/blob/94fcb6dedbc246da122576e429a47a53f373a210/lib/goth/token.ex#L78

Yes, my bad! 😅 

```elixir
{:ok, token} = Goth.Token.for_scope(client_email, @auth_scope)
```

Should be this - 

```elixir
{:ok, token} = Goth.Token.for_scope({client_email, @auth_scope})
```
---------------------------------------------------------------------------------------------

> Also, since it seems likely that a common use case will have only one Firebase account, forcing the user to supply the client_email parameter for all calls seems a bit unweildy. My inclination would be to make it an optional parameter

Yes makes sense, will make the `client_email ` optional 👍

**Comment by nitinstp23** [comment]
@bernardd I have made all the above changes. 

Please have a final look, then we can merge this PR!

File: lib/auth.ex [hunk 1]
```diff
@@ -7,48 +7,74 @@ defmodule FirebaseAdminEx.Auth do
   @doc """
   Get a user's info by UID
   """
-  @spec get_user(String.t()) :: tuple()
-  def get_user(uid), do: get_user(:localId, uid)
+  @spec get_user(String.t(), String.t()) :: tuple()
+  def get_user(uid, client_email \\ nil)
+  def get_user(uid, client_email), do: get_user(:localId, uid, client_email)
```
Comment (bernardd) [inline lib/auth.ex]: You don't need to have two `def get_user` lines here. You can just do

```elixir
def get_user(uid, client_email \\ nil), do: get_user(:localId, uid, client_emial)
```

Remember also that the `@spec` should now be
```elixir
@spec get_user(String.t(), String.t() | nil) :: tuple()
```

Same applies to the other functions below

**Comment by bernardd** [comment]
Other than my note above, it looks fine.

**Comment by nitinstp23** [comment]
Merged!

@bernardd Thanks for the review 👍

**Pull request merged**